bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/ui/navigator/CSymbolsContentProvider.java

```diff
@@ -90,24 +90,77 @@ public Object[] getChildren(Object parentElement) {
 
 	@Override
 	protected Object[] getTranslationUnitChildren(ITranslationUnit unit) {
+		if (!symbolsManager.isDirty(unit)) {
+			var symbols = symbolsManager.getTranslationUnitElements(unit);
+			if (symbols != null) {
+				return symbols;
+			}
+		}
 		if (loader != null) {
 			return loader.getChildren(unit);
 		}
 		return NO_CHILDREN;
 	}
 
+	/**
+	 * A variant of {@link DeferredTreeContentManager}. By adding a fixed {@link IDeferredWorkbenchAdapter}
+	 * we avoid to implement an adapter for {@link ITranslationUnit} to {@code IDeferredWorkbenchAdapter}.
+	 * This would also fail, because {@code cdt} has already a registered adapter from {@code ITranslationUnit}
+	 * to {@code IDeferredWorkbenchAdapter}.
+	 * It doesn't use a separate UI job to fill in the tree. With UI jobs, it's simply impossible
+	 * to know what has already been added when there are several loading jobs.
+	 * For our use case (load the whole symbols, then add it to the tree) a
+	 * {@link org.eclipse.swt.widgets.Display#syncExec(Runnable) syncExec()} is
+	 * sufficient.
+	 */
 	private static class DeferredCSymbolLoader extends DeferredTreeContentManager {
 		private final IDeferredWorkbenchAdapter adapter;
+		private final AbstractTreeViewer viewer;
 
 		public DeferredCSymbolLoader(AbstractTreeViewer viewer, IDeferredWorkbenchAdapter adapter) {
 			super(viewer);
+			this.viewer = viewer;
 			this.adapter = adapter;
 		}
 
 		@Override
 		protected IDeferredWorkbenchAdapter getAdapter(Object element) {
 			return adapter;
 		}
+
+		/**
+		 * Add child nodes, removing the error element if appropriate. Contrary
+		 * to the super implementation, this does <em>not</em> use a UI job but
+		 * a simple {@link org.eclipse.swt.widgets.Display#syncExec(Runnable)
+		 * syncExec()}.
+		 *
+		 * @param parent
+		 *            to add the {@code children} to
+		 * @param children
+		 *            to add to the {@code parent}
+		 * @param monitor
+		 *            is ignored
+		 */
+		@Override
+		protected void addChildren(Object parent, Object[] children, IProgressMonitor monitor) {
+			Control control = viewer.getControl();
+			if (control == null || control.isDisposed()) {
+				return;
+			}
+			control.getDisplay().syncExec(() -> {
```

Comment:
Does this need to be syncExec? syncExec (used to?) be subject to lock ups in Eclipse and should be avoided. Normally syncExec would be called because you want the result of the operation on the display thread. Alternative is asyncExec. The UIJob that the super class uses is asynchronous too (it does a Display.asyncExec in the UIJob's run)

Reply:
Besides this it LGTM

Reply:
> Does this need to be syncExec?

Good point. No I don't think so. I'll change it to asyncExec